src/Cms/PageCopy.php

```diff
@@ -0,0 +1,229 @@
+<?php
+
+namespace Kirby\Cms;
+
+use Kirby\Uuid\Uuid;
+use Kirby\Uuid\Uuids;
+
+/**
+ * Normalizes a newly generated copy of a page,
+ * adapting page slugs, UUIDs etc.
+ * (for single as well as multilang setups)
+ *
+ * @package   Kirby Cms
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ * @since     5.0.0
+ * @internal
+ */
+class PageCopy
+{
+	public function __construct(
+		public Page $copy,
+		public Page|null $original = null,
+		public bool $files = false,
+		public bool $children = false,
+		public array $uuids = []
+	) {
+	}
+
+	/**
+	 * Returns the pages collection for children to adapt
+	 */
+	public function children(): Pages
+	{
+		return match ($this->children) {
+			true  => $this->copy->index(drafts: true),
+			false => new Pages()
+		};
+	}
+
+	/**
+	 * Converts UUIDs for copied pages,
+	 * replacing the old UUID with a newly generated one
+	 * for all newly generated pages and files
+	 */
+	public function convertUuids(Language|null $language): void
+	{
+		if (Uuids::enabled() === false) {
+			return;
+		}
+
+		if ($language instanceof Language && $language->isDefault() === false) {
+			return;
+		}
+
+		// store old UUID
+		$old = $this->copy->uuid()->toString();
+
+		// re-generate UUID for the page
+		$this->copy = $this->copy->save(
+			['uuid' => Uuid::generate()],
+			$language?->code()
+		);
+
+		// track UUID change
+		$this->uuids[$old] = $this->copy->uuid()->toString();
+
+
+		$this->convertFileUuids($language);
+		$this->convertChildrenUuids($language);
+	}
+
+	/**
+	 * Re-generate UUIDs for each child recursively
+	 * and merge with the tracked changed UUIDs
+	 */
+	protected function convertChildrenUuids(Language|null $language): void
+	{
+		// re-generate UUIDs and track changes
+		foreach ($this->children() as $child) {
+			// always adapt files of subpages as they are
+			// currently always copied; but don't adapt
+			// children because we already operate on the index
+			$child = new PageCopy($child, files: true, uuids: $this->uuids);
+			$child->convertUuids($language);
+			$this->uuids = [...$this->uuids, ...$child->uuids];
+		}
+
+		// if children have not been copied over,
+		// track all children UUIDs from original page to
+		// remove/replace with empty string
+		if ($this->children === false) {
```

Comment:
I stumbled across the property name here while reading, because I thought it was somehow related to the children method. This is of course only happening because we use the children property in the Page method as a cache for the children method. But maybe it would help to rename the children and file properties to something like includeChildren and includeFiles or copyChildren and copyFiles. It's just a thought though after stumbling across it. It's not super important.

Reply:
I went with `withChildren` and `withFiles`.